compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -266,15 +278,238 @@ static LogicalResult replaceAllStoresWithTiledVersion(
+  SmallVector<int64_t> tileSizes;
+
+  for (auto val : requestedTileSizes) {
+    if (auto constIndexOp = val.getDefiningOp<arith::ConstantIndexOp>()) {
+      auto v = constIndexOp.value();
+      // When tiling is not defined, let's use 1 here.
+      tileSizes.push_back(v == 0 ? 1 : v);
+    } else {
+      return failure();
+    }
+  }
+  return tileSizes;
+}
+
+static SmallVector<AffineMinOp> collectTileSizeMinOps(
+    ArrayRef<OpFoldResult> tileSizes) {
+  SmallVector<AffineMinOp> res;
+
+  for (auto ofr : tileSizes) {
+    if (auto value = ofr.dyn_cast<Value>()) {
+      if (auto minOp = value.getDefiningOp<AffineMinOp>()) {
+        res.push_back(minOp);
+      } else {
+        res.push_back(AffineMinOp());
+      }
+    } else {
+      res.push_back(AffineMinOp());
+    }
+  }
+  return res;
+}
+
+static SmallVector<scf::ForOp> collectNestedLoops(scf::ForOp outerLoop) {
+  SmallVector<scf::ForOp> loops;
+
+  outerLoop.getOperation()->walk<WalkOrder::PreOrder>(
+      [&](scf::ForOp op) { loops.push_back(op); });
+
+  return loops;
+}
+
+// Specialize the tiled distribution loops with the main tile sizes.
+//
+// Inputs:
+//   - TilingResult including main distribution loops and other info
+//
+// Transformed output
+//   cond = (group_id_y != last) && (group_id_x != last)
+//   scf.if cond
+//     distribution loops with static shapes with the tile size
+//   else
+//     distribution loops with dynamic shapes with the tile size
+//
+// Steps:
+//   1. bail out if the loop bounds are already a multiple of the tile size.
+//   2. create a condition for scf.if
+//   3. clone the nested loops in the else block.
+//   4. update the bounded size ops of the original loop nest with the constant
+//      tile size.
+//   5. clone the updated loop nest in the then block.
+static LogicalResult specializeDistributionLoops(TilingResult &tilingResult,
+                                                 PatternRewriter &rewriter) {
+  auto &distLoops = tilingResult.loops;
+  auto tileSizes = getTileSizesAsInt64(tilingResult.tileSizesForLoops);
+  if (failed(tileSizes)) return failure();
+
+  assert(tileSizes->size() == distLoops.size());
+
+  // Check the eligilbility. Unsupported cases are:
+  //   1. Dynamic cases
+  //   2. UB < Tile size
+  //   3. UB == a multiple of the tile size
+  bool isAlreadyMultiple = true;
+  for (auto it : zip(distLoops, *tileSizes)) {
+    auto loop = std::get<0>(it);
+    auto ubOp = loop.getUpperBound().getDefiningOp<arith::ConstantIndexOp>();
+    if (!ubOp) {
+      // TODO: handle dynamic cases by adding more code to check the
+      // conditions.
+      return success();
+    }
+    auto ub = ubOp.value();
+    auto tileSize = std::get<1>(it);
+    if (ub % tileSize != 0) {
+      isAlreadyMultiple = false;
+    }
+    if (ub < tileSize) {
+      return success();
+    }
+  }
+
+  if (isAlreadyMultiple) return success();
+
+  auto minSizeOps = collectTileSizeMinOps(tilingResult.boundedSizesForLoops);
```

Comment:
There seems to be an assumption here on what `affine.min` operations are generated. Its not clear to me what those assumptions are. That might make this more fragile. Could you clarify?

Reply:
Yeah, this is unfortunately tied to the implementation detail, but I put the effort not to do a pattern matching from the generated scf.for loops. That's why I extended TilingResult to keep the bounded size information. If the structure is changed, we need to update the code too. Otherwise, there will be a compilation error. I think it is better than a pure pattern matching, which may fail silently.

A better way is to have an dialect op to abstract out the tiling operation. I don't have a good estimate for the implementation at the moment. I thought about using scf.foreach_thread, but it is the result of the workgroup distribution, but not the operation that dictates the tiling.

Let me know if you have a good idea to improve.

Reply:
Dialect op to abstract out the tiling operation has been tried before and abandoned.  So I would not go down that path. I dont follow the comments regarding use of `scf.foreach_thread`.

Reply:
The bounded size, represented as an affine.min op, is the key IR we need to find and replace with the constant tile size. `scf.foreach_thread` itself does not tie this information in the op. The min op is generated during the tiling process. Once we have generated `scf.foreach_thread`, we need to find the info from it, so for the current transformation, it does not help much.

Reply:
To echo from other comment, the `affine.min` op is an artifact. I dont think we should index on that. What you are really looking for is if `(upper_bound - lower_bound) % tileSize` is zero. Thats purely dependent on the loop bounds. The `affine.min` is part of the loop body (which you dont really need to look at).

Reply:
I can use that for the condition part, but the bounded size (represented as affine.min currently) used in the body of the nested loop is what I want to replace with the tile size.